### PR TITLE
RichText Trigger onChange on input

### PR DESCRIPTION
**Pull request by youknowriad**
This will be necessary to achieve #4951 

It also fixes some issues while the save draft link doesn't show up until we focus out of the component.
It may introduce a difference in a way undo/redo works but I believe we should handle those consistently (it's not the case right now, RichText triggers and undo state per word while PlainText it's per character). I'm thinking we could use the `withHistory` a Reducer enhancer to group similar changes into one undo level.

**Testing instructions**

 - Test several writing flow interactions

### CODE REVIEW

**Comment by aduth** [comment]
See: #2758

(Particularly bits about undo history)

**Comment by iseulde** [comment]
I'm currently rebasing a WIP branch that does the undo signalling with a buffer in the history. I'm also leaving out selection now since it would make it more complex. I feel like we should fix undo together this PR.

File: blocks/rich-text/index.js [hunk 1]
```diff
@@ -680,19 +670,10 @@ export default class RichText extends Component {
 		this.savedContent = this.props.value;
 		this.setContent( this.savedContent );
 		this.editor.selection.moveToBookmark( bookmark );
-
-		// Saving the editor on updates avoid unecessary onChanges calls
-		// These calls can make the focus jump
-		this.editor.save();
 	}
 
 	setContent( content ) {
```
Comment (iseulde) [inline blocks/rich-text/index.js]: I forgot to add `content = ''` here 🙈

**Comment by youknowriad** [comment]
So from my testing, performance is good enough here but I'd appreciate that other people try it to confirm

File: blocks/rich-text/index.js [hunk 2]
```diff
@@ -92,6 +93,7 @@ export default class RichText extends Component {
 		this.getSettings = this.getSettings.bind( this );
 		this.onSetup = this.onSetup.bind( this );
 		this.onChange = this.onChange.bind( this );
+		this.throttledOnChange = throttle( this.onChange.bind( this ), 500 );
```
Comment (aduth) [inline blocks/rich-text/index.js]: If we can, I'd love if we would avoid throttling. Case in point: We're not cancelling a scheduled change when the component unmounts.

Comment (youknowriad) [reply blocks/rich-text/index.js]: Yep, if the performance fix is enough, I'm fine dropping it.

**Comment by aduth** [comment]
@youknowriad I created a branch off of this one which includes an iteration on `rememo` to improve per-key selector caching.

https://github.com/WordPress/gutenberg/tree/try/weak-map-cache

**Comment by youknowriad** [comment]
@aduth Awesome, I'm not seeing any performance issues any more 👍 

---
On the performance side, maybe instead of using `getBlock` selector, we could target block properties directly `getBlockName`, `getBlockAttributes`, `isValidBlock` etc... These granular properties are less likely to change compared to the block object which will avoid unnecessary renders in `connect` calls.

**Comment by youknowriad** [comment]
Rebase this with the changes from #5002 it seems better now.

File: blocks/rich-text/index.js (hunk 2 above)
Comment (youknowriad) [reply blocks/rich-text/index.js]: I tried dropping it but I think performance-wise it's slightly better if we keep it

**Comment by gziolo** [comment]
@youknowriad, see https://github.com/WordPress/gutenberg/pull/5002/files#r167532619, can we bring the polyfill as part of this PR?

File: blocks/rich-text/index.js [hunk 3]
```diff
@@ -692,19 +684,10 @@ export class RichText extends Component {
...
 	}
 
-	setContent( content ) {
-		if ( ! content ) {
-			content = '';
-		}
-
-		content = renderToString( content );
-		this.editor.setContent( content );
+	setContent( content = '' ) {
```
Comment (gziolo) [inline blocks/rich-text/index.js]: Nice refactoring :)

File: blocks/rich-text/index.js [hunk 4]
```diff
@@ -366,21 +369,12 @@ export class RichText extends Component {
...
 	/**
 	 * Handles any case where the content of the tinyMCE instance has changed.
 	 */
 	onChange() {
-		// Note that due to efficiency, speed and low cost requirements isDirty may
-		// not reflect reality for a brief period immediately after a change.
-		if ( this.editor.isDirty() ) {
-			this.fireChange();
-		}
+		this.savedContent = this.state.empty ? [] : this.getContent();
```
Comment (gziolo) [inline blocks/rich-text/index.js]: @aduth - should we extract `[]` to its own shared variable to avoid having always one reference to compare? I wanted to double check - you know better all those perf bits. It might not make any sense in this context :)

File: blocks/rich-text/index.js [hunk 5]
```diff
@@ -713,21 +696,20 @@ export class RichText extends Component {
...
 	componentDidUpdate( prevProps ) {
 		// The `savedContent` var allows us to avoid updating the content right after an `onChange` call
 		if (
+			!! this.editor &&
 			this.props.tagName === prevProps.tagName &&
 			this.props.value !== prevProps.value &&
-			this.props.value !== this.savedContent &&
-			! isEqual( this.props.value, prevProps.value ) &&
-			! isEqual( this.props.value, this.savedContent )
+			this.props.value !== this.savedContent
```
Comment (gziolo) [inline blocks/rich-text/index.js]: It seems like shallow compare is enough here.

**Comment by gziolo** [comment]
I played with `cmd + Z` and noticed a bit strange behavior, the following screencast contains more details shared as I type:
on-change

File: blocks/rich-text/index.js (hunk 2 above)
Comment (gziolo) [reply blocks/rich-text/index.js]: It seems like it always dispatched change when I type the first character. Should we wait until 500ms passes instead?

**Comment by gziolo** [comment]
I also recorded another screencast where you can observe how strangely undo behaves:

on-change

I'm doing `cmd + Z` until undo history gets emptied.

File: blocks/rich-text/index.js (hunk 2 above)
Comment (iseulde) [reply blocks/rich-text/index.js]: @youknowriad What can be do to get rid of this? :)

**Comment by gziolo** [comment]
I'm still able to type something and close the tab without seeing the prompt to save your changes for my draft:

no-changes-detected

**Comment by youknowriad** [comment]
@gziolo 

 - I can't seem to reproduce these undo issues though granted that it's not perfect, it's not perfect in master and as I said let's explore those separately in #4956 and #4959 

 - About the last screenshot, this is due to the throttle timeout. I think it's fine for now, I don't expect people to close their tab less than 0.5 seconds after editing the post unless they don't want to save these changes :). I'm also hopeful we could get rid of the throttle entirely at some point when we improve rerendering performance event more.

**Comment by aduth** [comment]
I'd be inclined to omitting the `throttle` from the start to force our hand in addressing the performance issues. 😅 

I'll try to take a pass at these today (non-blocking).

**Review by gziolo — APPROVED** [review]
I tested in enough. Let's merge it and continue in the follow-up PRs as discussed 👍

File: blocks/rich-text/index.js [hunk 6]
```diff
@@ -713,21 +704,20 @@ export class RichText extends Component {
...
 		) {
 			this.updateContent();
 		}
 	}
-
```
Comment (iseulde) [inline blocks/rich-text/index.js]: Removed line break by accident?

File: blocks/rich-text/index.js [hunk 7]
```diff
@@ -508,8 +506,6 @@ export class RichText extends Component {
 				return;
 			}
 
-			this.fireChange();
```
Comment (iseulde) [inline blocks/rich-text/index.js]: Why is this removed?

**Comment by youknowriad** [comment]
I don't know, this was part of one of your commits, I can restore :)

**Comment by iseulde** [comment]
@youknowriad I think I removed that there assuming changes would be in sync all the time, but with the throttle that wouldn't be the case?

**Comment by aduth** [comment]
Noting that in the course of writing a post with a single paragraph containing only 32 words, I accumulated 71 copies of editor state via undo history (and that this will only grow over the course of writing a post in the same session).

**Comment by aduth** [comment]
Something revealed by this change: Should we be _debouncing_ autosave instead of throttling? It feels as though it saves much more frequently now, since it can occur many times while typing within the same paragraph block.

Debouncing isn't great either, since just because I don't stop typing for X amount of time doesn't mean the post should never save during that time.

Or maybe the autosave should just be made less frequent than every 10 seconds?

File: blocks/rich-text/index.js [hunk 8]
```diff
@@ -149,6 +151,7 @@ export class RichText extends Component {
 		editor.on( 'BeforeExecCommand', this.maybePropagateUndo );
```
Comment (aduth) [inline blocks/rich-text/index.js]: Did we test to see how TinyMCE's undo history interacts with our own?

- Are we calling `onChange` with the correct value after TinyMCE finishes undo?
- Should we be bypassing TinyMCE's undo altogether (also preventing it from accumulating any history)?

Comment (iseulde) [reply blocks/rich-text/index.js]: @aduth I'm proposing these changes in #4956. Good point about preventing TinyMCE accumulating any history at all. We just need the signals, not the store.

**Comment by iseulde** [comment]
@aduth I don't mind if we save frequently, but it is quite visually distracting ATM. Sounds like it need to be more subtle, similar to Google Docs. We also have to make sure we don't just pile these changes up as revisions, instead we should maybe reuse the last revision and only create a new revision on a user action or at least way less often (each revision is a post in the DB).

**Comment by aduth** [comment]
>We also have to make sure we don't just pile these changes up as revisions

Yeah, this is an ongoing issue unfortunately: #3656
